### PR TITLE
feat(ddm): Add order by support for the new metrics API

### LINKED ISSUE #60785: Add support for order by
Add ability to pass order and limit to /metrics/data endpoint.

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds support for order by and limit in the layer. It's important to note that the support, is for a single order by only one of the aggregates in the query. This limitation is given by the metrics layer, which supports ordering in totals query and only on the aggregate value.

The engine for executing queries works in the following way (suppose you ask for `transaction`, `sum(x)`, `sum(y)` and you order by `sum(x)`):
1. The reference query is found by looking at the query whose `field` is `sum(x)`. A reference query is a query that is used by the following queries to align to, in order to maintain the sort order.
1. A totals query for `transaction`, `sum(x)` is executed.
2. The groups returned by that query are applied on the series query for `transaction`, `sum(x)`.
3. The reference query series component is aligned on the totals, since the layer doesn't allow ordering on series. Now the state is that we have a reference query with both totals and series ordered correctly (based on `sum(x)`).
4. The groups returned by totals reference query are applied as filters on the following queries, in this case the query `transaction`, `sum(y)` will be executed.
5. Once the query is executed, the values will be aligned to the ones of the reference query, effectively resulting in two lists were the `transaction` is used to define order (because it's the only common value between the two queries).
6. The values of the queries are now added together and returned as a list of `QueryResult` that are going to be stiched together by the translation layer.

In conclusion, a query with `n` aggregates, will result in `n * 2` queries. If you also consider the dynamic interval calculation in place, we could reach a maximum of `(2 * 2) + (n - 1) * 2` since we will have to first run at most 2 times the totals and series query until we find a good interval bigger than the optimal.

Closes https://github.com/getsentry/sentry/issues/60785

### CODE REVIEW

File: tests/sentry/sentry_metrics/querying/test_api.py [hunk 1]
```diff
@@ -28,11 +28,11 @@ def setUp(self):
 
         for value, transaction, platform, env, time in (
             (1, "/hello", "android", "prod", self.now()),
-            (3, "/hello", "ios", "dev", self.now()),
+            (6, "/hello", "ios", "dev", self.now()),
```
Comment (iambriccardo) [inline tests/sentry/sentry_metrics/querying/test_api.py:31]: Changed the values to make the order by always with different values (otherwise the test execution time could lead to a non-deterministic ordering of the same valued rows).

File: src/sentry/sentry_metrics/querying/api.py [hunk 2]
```diff
@@ -672,13 +932,20 @@ def run_metrics_query(
...
-            .set_rollup(Rollup(interval=interval))
-            .set_limit(SNUBA_QUERY_LIMIT)
+    for field, timeseries in parser.generate_queries(environments=environments):
+        query = base_query.set_query(timeseries).set_rollup(Rollup(interval=interval))
+
+        # We will apply the order by if it only matches the field. This is done since for now we don't support a custom
+        # since for order bys.
+        query_order_by = None
+        if order_by and field == order_by.removeprefix("-"):
+            query_order_by = order_by
```
Comment (iambriccardo) [inline src/sentry/sentry_metrics/querying/api.py:942]: This is not very nice, but it's the simplest solution especially since we are not sure where to head w.r.t. the API with multiple queries.

File: src/sentry/sentry_metrics/querying/api.py [hunk 3]
```diff
@@ -605,27 +858,32 @@ def _group(values, block):
...
+    # TODO: we might want to give users the ability to just run totals queries. In that case, we will have to build
+    #   the logic that is able to not require an interval.
     assert start is not None and end is not None and interval is not None
 
     # We build the intervals that we will return to the API user.
     intervals = _build_intervals(start, end, interval)
 
     translated_groups = []
-    for group_key, group_metrics in sorted(intermediate_groups.items(), key=lambda v: v[0]):
+    for group_key, group_metrics in intermediate_groups.items():
```
Comment (iambriccardo) [inline src/sentry/sentry_metrics/querying/api.py:869]: Since we use ordered dictionaries, we do not need to sort the groups.

File: src/sentry/sentry_metrics/querying/api.py [hunk 4]
```diff
@@ -553,7 +808,7 @@ def _translate_query_results(execution_results: List[QueryResult]) -> Mapping[st
     interval: Optional[int] = None
 
     # For efficiency reasons, we translate the incoming data into our custom in-memory representations.
-    intermediate_groups: Dict[GroupKey, Dict[str, GroupValue]] = {}
+    intermediate_groups: OrderedDict[GroupKey, OrderedDict[str, GroupValue]] = OrderedDict()
```
Comment (iambriccardo) [inline src/sentry/sentry_metrics/querying/api.py:811]: We use an ordered dictionary, since we have to preserve the order of the values that are returned by the `QueryExecutor`.

**Review by obostjancic — APPROVED** [review]
Reviewed offline. LGTM
